cmd/lotus-shed/indexes.go

```diff
@@ -572,14 +639,148 @@ var inspectEventsCmd = &cli.Command{
 			prevTs = currTs
 			currTs, err = api.ChainGetTipSet(ctx, currTs.Parents())
 			if err != nil {
-				return fmt.Errorf("failed to load tipset %s: %w", currTs, err)
+				return xerrors.Errorf("failed to load tipset %s: %w", currTs, err)
 			}
 		}
 
 		return nil
 	},
 }
 
+// amtRootForEvents generates the events AMT root CID for a given message's events, and returns
+// whether the message has events, a string describing any non-fatal problem encountered,
+// and a fatal error if one occurred.
+func amtRootForEvents(
+	ctx context.Context,
+	api lapi.FullNode,
+	tsKeyCid cid.Cid,
+	prevTsKey types.TipSetKey,
+	stmtSelectEventIdAndEmitter, stmtSelectEventEntries *sql.Stmt,
+	message lapi.Message,
+	addrIdCache map[address.Address]abi.ActorID,
+) (cid.Cid, bool, string, error) {
+
+	events := make([]cbg.CBORMarshaler, 0)
+
+	rows, err := stmtSelectEventIdAndEmitter.Query(tsKeyCid.Bytes(), message.Cid.Bytes())
+	if err != nil {
+		return cid.Undef, false, "", xerrors.Errorf("failed to query events: %w", err)
+	}
+	defer func() {
+		_ = rows.Close()
+	}()
+
+	for rows.Next() {
+		var eventId int
+		var emitterAddr []byte
+		if err := rows.Scan(&eventId, &emitterAddr); err != nil {
+			return cid.Undef, false, "", xerrors.Errorf("failed to scan row: %w", err)
+		}
+
+		addr, err := address.NewFromBytes(emitterAddr)
+		if err != nil {
+			return cid.Undef, false, "", xerrors.Errorf("failed to parse address: %w", err)
+		}
+		var actorId abi.ActorID
+		if id, ok := addrIdCache[addr]; ok {
+			actorId = id
+		} else {
+			if addr.Protocol() != address.ID {
+				// use the previous tipset (height+1) to do an address lookup because the actor
+				// may have been created in the current tipset (i.e. deferred execution means the
+				// changed state isn't available until the next epoch)
+				idAddr, err := api.StateLookupID(ctx, addr, prevTsKey)
+				if err != nil {
+					// TODO: fix this? we should be able to resolve all addresses
+					return cid.Undef, false, fmt.Sprintf("failed to resolve address (%s), could not compare amt", addr.String()), nil
+				}
+				addr = idAddr
+			}
+			id, err := address.IDFromAddress(addr)
+			if err != nil {
+				return cid.Undef, false, "", xerrors.Errorf("failed to get ID from address: %w", err)
+			}
+			actorId = abi.ActorID(id)
+			addrIdCache[addr] = actorId
+		}
+
+		event := types.Event{
+			Emitter: actorId,
+			Entries: make([]types.EventEntry, 0),
+		}
+
+		rows2, err := stmtSelectEventEntries.Query(eventId)
+		if err != nil {
+			return cid.Undef, false, "", xerrors.Errorf("failed to query event entries: %w", err)
+		}
+		defer func() {
+			_ = rows2.Close()
+		}()
+
+		for rows2.Next() {
+			var flags []byte
+			var key string
+			var codec uint64
+			var value []byte
+			if err := rows2.Scan(&flags, &key, &codec, &value); err != nil {
+				return cid.Undef, false, "", xerrors.Errorf("failed to scan row: %w", err)
+			}
+			entry := types.EventEntry{
+				Flags: flags[0],
```

Comment:
Why do we only read the first byte of the flag here ?

Reply:
It's just a `uint8`, so we only need the first byte: https://github.com/filecoin-project/lotus/blob/d4b8af99b201053a7ae40397052fe75522c867ac/chain/types/event.go#L23

I'm not sure why we store it as a `BLOB` though, maybe it was imagined it would be extended at some point? I doubt it.